File: backend/src/kb_backend/resolve.py
"""Shared resolution engine (docs/PRD.md §4.6.1) — the single implementation
behind both the list-filter endpoint and the single-point resolve endpoint
(issue #5). Ported from frontend-mock/assets/app.js's `resolveAnswer` /
`liveGroups` / `coordCompatible`, with PRD text (not just the JS) as the
source of truth for the created_at tie-breaks — see
docs/specs/2026-08-08-resolve-engine-design.md.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from typing import Any, Literal

from sqlalchemy import select
from sqlalchemy.orm import Session

from .models.answer import Answer
from .models.dimension import DimensionDefinition

ResolveStatus = Literal["exact", "weighted", "default", "fallback-latest", "none"]


@dataclass(frozen=True)
class LiveGroup:
    coord: dict[str, Any]
    coord_hash: str
    live_answer: Answer
    spec: int
    weight: int


@dataclass(frozen=True)
class ResolveResult:
    status: ResolveStatus
    answer: Answer | None


@dataclass(frozen=True)
class AnswerGroupSummary:
    coord: dict[str, Any]
    coord_hash: str
    revoked: bool
    version_count: int
    latest_answer: Answer
    live_answer: Answer | None


def _dimension_weights(db: Session, keys: set[str]) -> dict[str, int]:
    if not keys:
        return {}
    rows = db.execute(
        select(DimensionDefinition.key, DimensionDefinition.weight).where(DimensionDefinition.key.in_(keys))
    ).all()
    return dict(rows)


def compute_live_groups(db: Session, kb_id: int, kp_id: int, at: date) -> list[LiveGroup]:
    """Step 1 of §4.6.1: group this knowledge point's non-revoked answers by
    coord_hash, and within each group pick the "current version" at time
    `at` — the row with the largest (effective_time, created_at), among
    rows with effective_time <= at. A group with no such row (every version
    is later than `at`) does not exist at this point in time and is
    dropped."""
    rows = (
        db.execute(
            select(Answer).where(
                Answer.knowledge_base_id == kb_id,
                Answer.knowledge_point_id == kp_id,
                Answer.revoked.is_(False),
                Answer.effective_time <= at,
            )
        )
        .scalars()
        .all()
    )

    by_hash: dict[str, list[Answer]] = {}
    for row in rows:
        by_hash.setdefault(row.coord_hash, []).append(row)

    all_keys: set[str] = set()
    for versions in by_hash.values():
        all_keys.update(versions[0].coord.keys())
    weights = _dimension_weights(db, all_keys)

    groups = []
    for coord_hash, versions in by_hash.items():
        # id as a final tie-break: `versions` comes from an unordered SELECT
        # (no ORDER BY), so relying on Python's stable max() over
        # MySQL's execution-plan-dependent row order would make the winner
        # non-deterministic whenever (effective_time, created_at) also tie.
        # Found by the Kimi review gate on PR #21.
        current = max(versions, key=lambda a: (a.effective_time, a.created_at, a.id))
        coord = current.coord
        groups.append(
            LiveGroup(
                coord=coord,
                coord_hash=coord_hash,
                live_answer=current,
                spec=len(coord),
                weight=sum(weights.get(k, 0) for k in coord),
            )
        )
    return groups


def compute_all_answer_groups(db: Session, kb_id: int, kp_id: int, at: date) -> list[AnswerGroupSummary]:
    """For the read-only answer-group tree (issue #7 §2) — deliberately
    separate from `compute_live_groups`, not a wrapper around it:
    `compute_live_groups` filters `revoked.is_(False)` at the SQL level, so
    a whole-chain-revoked coord group (e.g. from edit_answer's migration
    branch) never appears in its output at all. The tree view must show
    revoked chains too (struck through), per PRD §4.4's "全部答案的分组树"
    and frontend-mock's kpAnswerTree — so this queries every answer for the
    knowledge point, revoked or not, effective yet or not."""
    rows = (
        db.execute(
            select(Answer).where(
                Answer.knowledge_base_id == kb_id,
                Answer.knowledge_point_id == kp_id,
            )
        )
        .scalars()
        .all()
    )

    by_hash: dict[str, list[Answer]] = {}
    for row in rows:
        by_hash.setdefault(row.coord_hash, []).append(row)

    summaries = []
    for coord_hash, versions in by_hash.items():
        latest = max(versions, key=lambda a: (a.effective_time, a.created_at, a.id))
        # Revocation is a whole-chain property (PRD §6 rule #4) — every row
        # sharing a coord_hash carries the same `revoked` value, so the
        # latest version's flag speaks for the whole chain.
        live = next(
            (
                v
                for v in sorted(versions, key=lambda a: (a.effective_time, a.created_at, a.id), reverse=True)
                if not v.revoked and v.effective_time <= at
            ),
            None,
        )
        summaries.append(
            AnswerGroupSummary(
                coord=latest.coord,
                coord_hash=coord_hash,
                revoked=latest.revoked,
                version_count=len(versions),
                latest_answer=latest,
                live_answer=live,
            )
        )
    return sorted(summaries, key=lambda s: s.coord_hash)


def _coord_compatible(group_coord: dict[str, Any], query_coord: dict[str, Any]) -> bool:
    # Every key the GROUP specifies must agree with the query wherever the
    # query also specifies it; a key the query asks about but the group
    # never wrote is not checked (and vice versa). A coord={} group is
    # therefore always compatible — it has nothing to disagree on — which is
    # a deliberate, PRD/demo-sanctioned property, not an oversight: it lets
    # the default answer act as a low-priority (spec=0) fallback candidate
    # rather than being excluded outright.
    for key, value in group_coord.items():
        if key not in query_coord:
            continue
        if value != query_coord[key]:
            return False
    return True


def resolve(groups: list[LiveGroup], query_coord: dict[str, Any]) -> ResolveResult:
    """Step 2 of §4.6.1. `query_coord` must already be normalized (see
    coord.normalize_coord) so its values compare equal to the groups'
    already-normalized coord values."""
    if not groups:
        return ResolveResult(status="none", answer=None)

    if not query_coord:
        default_group = next((g for g in groups if g.spec == 0), None)
        if default_group is not None:
            return ResolveResult(status="default", answer=default_group.live_answer)
        latest = max(
            groups, key=lambda g: (g.live_answer.effective_time, g.live_answer.created_at, g.live_answer.id)
        )
        return ResolveResult(status="fallback-latest", answer=latest.live_answer)

    candidates = [g for g in groups if _coord_compatible(g.coord, query_coord)]
    if not candidates:
        return ResolveResult(status="none", answer=None)

    top = max(
        candidates,
        key=lambda g: (
            g.spec,
            g.weight,
            g.live_answer.effective_time,
            g.live_answer.created_at,
            g.live_answer.id,
        ),
    )
    is_exact = top.spec == len(query_coord) and all(
        top.coord[k] == query_coord.get(k) for k in top.coord
    )
    return ResolveResult(status="exact" if is_exact else "weighted", answer=top.live_answer)

File: frontend/src/utils/resolve.ts
import type { Filters } from './dimension'
import type { Dimension } from '@/api/dimension'
/**
 * 「当前答案」tab 的前端复算逻辑——必须与后端 resolve.py 逐字段一致，
 * 否则该 tab 的「此条件下生效」标记会与 /resolve、列表页 resolved 预览
 * 对同一知识点给出不同答案。迁移自 React 版 api/answers.ts。
 */
import type { AnswerGroup } from '@/api/knowledgePoint'
import { coordValueEquals } from './coord'

function coordSpec(coord: Record<string, unknown>): number {
  return Object.keys(coord).length
}

function coordWeight(coord: Record<string, unknown>, dimensions: Dimension[]): number {
  return Object.keys(coord).reduce((sum, key) => sum + (dimensions.find(d => d.key === key)?.weight ?? 0), 0)
}

/**
 * PRD §4.6.1 规则 3 第一条——精确镜像后端 resolve.py::_coord_compatible：
 * 组自身写了的每个 key，凡查询也指定了的，取值必须一致；查询问了而组没写
 * （或反之）的 key 不参与检查，因此 coord={} 永远兼容。
 * 比较必须走 coordValueEquals 而非裸 String()/==（Codex 结论，PR #24）：
 * number 型筛选值是保精度字符串而组内值是 JSON 数字，朴素比较会拒绝
 * 组 `1` vs 筛选 `"1.0"` 这类等价值，静默藏掉 /resolve 本会命中的答案。
 */
function coordCompatible(groupCoord: Record<string, unknown>, query: Filters, dimensions: Dimension[]): boolean {
  for (const [key, value] of Object.entries(groupCoord)) {
    if (!(key in query))
      continue
    const fieldType = dimensions.find(d => d.key === key)?.field_type
    if (!coordValueEquals(fieldType, value, query[key]))
      return false
  }
  return true
}

/**
 * 复刻 resolve.py::resolve() 真实的 5 键排序元组——不是 PRD §4.6.1 文字描述的
 * 3 键版本。effective_time 是天粒度、平局常见；created_at/id 是服务端为保证
 * 平局裁决确定性专门加的（resolve.py:88-93，PR #21 评审）。只用 3 键会让这里
 * 的排序与服务端对同一知识点算出的结果不一致（设计文档 §4.1）。
 *
 * 已知且接受的偏差（设计文档 §4.1）：coord key 已不在 `dimensions` 里（全局
 * 停用或本库禁用）时权重按 0 计，而后端自己的权重查询不过滤状态。修复需要
 * 新的后端接口暴露停用维度权重，超出范围，不在此处处理。
 */
export function sortLiveGroupsByPriority(
  groups: AnswerGroup[],
  filters: Filters,
  dimensions: Dimension[],
): AnswerGroup[] {
  const live = groups.filter(g => g.live_answer !== null)
  const compatible
    = Object.keys(filters).length === 0 ? live : live.filter(g => coordCompatible(g.coord, filters, dimensions))
  return [...compatible].sort((a, b) => {
    const specDiff = coordSpec(b.coord) - coordSpec(a.coord)
    if (specDiff !== 0)
      return specDiff
    const weightDiff = coordWeight(b.coord, dimensions) - coordWeight(a.coord, dimensions)
    if (weightDiff !== 0)
      return weightDiff
    const av = a.live_answer!
    const bv = b.live_answer!
    if (av.effective_time !== bv.effective_time)
      return av.effective_time < bv.effective_time ? 1 : -1
    if (av.created_at !== bv.created_at)
      return av.created_at < bv.created_at ? 1 : -1
    return bv.id - av.id
  })
}

/**
 * 「此条件下生效」标记：只有存在筛选条件可供消歧时才有意义，且第一名不能与
 * 第二名在 spec（条件个数）上平局——平局说明「赢家」是靠权重/时间选出来的，
 * 不是该具体度下唯一兼容的候选。
 */
export function hasUniqueTopMatch(sortedGroups: AnswerGroup[], hasFilter: boolean): boolean {
  if (!hasFilter || sortedGroups.length === 0)
    return false
  if (sortedGroups.length === 1)
    return true
  return coordSpec(sortedGroups[0].coord) > coordSpec(sortedGroups[1].coord)
}
